// modloader64_api/ModLoaderAPIInjector.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.setupMLInjects = exports.ModLoaderConstructorInjector = exports.ModLoaderConstructorBus = exports.ModLoaderAPIInject = void 0;
const EventHandler_1 = require("./EventHandler");
function ModLoaderAPIInject() {
    return function (target, key) {
        if (target.ModLoader === undefined) {
            target['ModLoader'] = {};
        }
        if (target.ModLoader.ModLoaderAPIInject === undefined) {
            target.ModLoader['ModLoaderAPIInject'] = {};
        }
        if (target.ModLoader.ModLoaderAPIInject.Targets === undefined) {
            target.ModLoader.ModLoaderAPIInject['Targets'] = new Map();
        }
        target.ModLoader.ModLoaderAPIInject.Targets.set(key, "yes");
    };
}
exports.ModLoaderAPIInject = ModLoaderAPIInject;
exports.ModLoaderConstructorBus = new EventHandler_1.EventBus();
function ModLoaderConstructorInjector(target) {
    var original = target;
    var f = function (...args) {
        let inst = new original(...args);
        exports.ModLoaderConstructorBus.emit('give', inst);
        return inst;
    };
    f.prototype = original.prototype;
    return f;
}
exports.ModLoaderConstructorInjector = ModLoaderConstructorInjector;
function setupMLInjects(instance, api) {
    let p = Object.getPrototypeOf(instance);
    if (p.hasOwnProperty('ModLoader')) {
        if (p.ModLoader.hasOwnProperty("hasBeenProcessed")) {
            return;
        }
        if (p.ModLoader.hasOwnProperty('ModLoaderAPIInject')) {
            if (p.ModLoader.ModLoaderAPIInject.hasOwnProperty("Targets")) {
                p.ModLoader.ModLoaderAPIInject.Targets.forEach((value, key) => {
                    instance[key] = api;
                });
            }
        }
    }
}
exports.setupMLInjects = setupMLInjects;
